File: examples/node/rag_demo.js
const fs = require('fs');
const {
    Chain,
    PromptTemplate,
    SambaNovaLLM,
    TextLoader,
    InMemoryVectorStore,
    MockEmbeddings,
    Document
} = require('../../node/index');

// Usage: node rag_demo.js

async function main() {
    console.log("--- RAG Demo ---");

    // 1. Setup Data
    const filename = "node_knowledge.txt";
    fs.writeFileSync(filename, `
  Node.js bindings for Mini LangChain use napi-rs.
  They provide a Promise-based API for async operations.
  You can use standard JS classes like Chain and PromptTemplate.
  `);

    // 2. Load
    const loader = new TextLoader(filename);
    const docs = loader.load();
    console.log(`Loaded ${docs.length} documents.`);

    // 3. Index
    const vectorStore = new InMemoryVectorStore(new MockEmbeddings());
    // Note: addDocuments might be async depending on implementation
    await vectorStore.addDocuments(docs);

    // 4. Retrieve
    const query = "How are Node.js bindings implemented?";
    const results = await vectorStore.similaritySearch(query, 1);
    const context = results[0].pageContent;
    console.log(`\nRetrieved Context: ${context.trim()}`);

    // 5. Generate
    const prompt = new PromptTemplate(
        "Context: {context}\n\nQuestion: {query}\n\nAnswer:"
    );
    const llm = new SambaNovaLLM("Meta-Llama-3.1-8B-Instruct");
    const chain = new Chain(prompt, llm);

    const answer = await chain.invoke({ context, query });
    console.log("\nAnswer:");
    console.log(answer);

    // Cleanup
    fs.unlinkSync(filename);
}

main();

File: examples/python/rag_demo.py
import asyncio
import os
from mini_langchain import (
    Chain, 
    PromptTemplate, 
    OpenAI, 
    SambaNovaLLM,
    TextLoader,
    InMemoryVectorStore,
    MockEmbeddings
)

async def main():
    print("--- RAG Demo ---")
    
    # 1. Create a dummy knowledge base file
    with open("knowledge.txt", "w") as f:
        f.write("Mini LangChain is a high-performance LLM orchestration library.\n")
        f.write("It is written in Rust and provides bindings for Python and Node.js.\n")
        f.write("It supports many providers including SambaNova, OpenAI, and Anthropic.\n")

    # 2. Load Documents
    loader = TextLoader("knowledge.txt")
    docs = loader.load()
    print(f"Loaded {len(docs)} documents.")

    # 3. Index Documents (using MockEmbeddings for demo purposes)
    # In a real app, use OpenAIEmbeddings or similar
    embeddings = MockEmbeddings() 
    vector_store = InMemoryVectorStore(embeddings)
    await vector_store.add_documents(docs)

    # 4. Retrieve context
    query = "What languages does Mini LangChain support?"
    results = await vector_store.similarity_search(query, k=1)
    context = results[0].page_content
    print(f"\nRetrieved Context: {context}")

    # 5. Generate Answer
    prompt = PromptTemplate(
        "Context: {context}\n\nQuestion: {question}\n\nAnswer:"
    )
    
    # Using SambaNova for generation
    llm = SambaNovaLLM(model="Meta-Llama-3.1-8B-Instruct")
    chain = Chain(prompt, llm)

    response = await chain.invoke({
        "context": context,
        "question": query
    })

    print("\nAnswer:")
    print(response)

    # Cleanup
    os.remove("knowledge.txt")

if __name__ == "__main__":
    asyncio.run(main())
